Add Cookstyle 5.2 release notes (#182)
chef/cookstyle: RELEASE_NOTES.md

## Cookstyle 5.2

### New Chef Cops

#### Chef/InvalidPlatformMetadata

> The `InvalidPlatformMetadata` cop detects invalid platform names being used in the supports metadata.rb method. It uses a hardcoded set of common typos so it's not able to detect all invalid platforms, but it does detect the most common mistakes found on the Supermarket.

Reviewer: ```suggestion
The `InvalidPlatformMetadata` cop detects invalid platform names in the supports metadata.rb method. It uses a hardcoded set of common typos so it can't detect all invalid platforms, but it does detect the most common mistakes found on the Supermarket.
```

`Enabled by default`: True

`Autocorrects`: True (deletes the occurrence)

#### Chef/RespondToResourceName

The `RespondToResourceName` cop detects the usage of respond_to?(:resource_name) in resources to provide backwards compatibility. This `respond_to?` check is no longer necessary in Chef Infra Client 12.5+

`Enabled by default`: True

`Autocorrects`: True (deletes the `if respond_to?` occurence)

#### Chef/RespondToProvides

The `RespondToProvides` cop detects the usage of respond_to?(:provides) in providers to provide backwards compatibility. This `respond_to?` check is no longer necessary in Chef Infra Client 12+

`Enabled by default`: True

`Autocorrects`: True (deletes the `if respond_to?` occurence)

#### Chef/RespondToInMetadata

> The `RespondToInMetadata` cop detects the usage of `if respond_to?(:foo)` to gate newer metadata methods from running on older Chef Infra Client releases. In Chef Infra Client 12.15 and later this is no longer necessary as chef-client will no longer fail if unknown metadata is encoutered.

Reviewer: ```suggestion
The `RespondToInMetadata` cop detects the usage of `if respond_to?(:foo)` to prevent newer metadata methods from running on Chef Infra Client versions older than 12.15. The chef-client doesn't fail when it encounters unknown metadata in Chef Infra Client versions 12.15.
```

`Enabled by default`: True

`Autocorrects`: True (deletes the `if respond_to?` occurence)

#### Chef/LongDescriptionMetadata

> The `LongDescriptionMetadata` cop detects usage of `long_description` in metadata.rb. The `long_description` metadata was never utilized by Chef Infra or Supermarket, but cookbooks commonly load their README.md file into the metadata. This increases the size of metadata stored on Chef Infra Server and provides no benefit to users.

Reviewer: ```suggestion
The `LongDescriptionMetadata` cop detects usage of `long_description` in metadata.rb. Chef Infra and Supermarket never utilized the `long_description` metadata, but cookbooks commonly load their README.md file into the metadata. This increases the size of metadata stored on Chef Infra Server without providing a clear benefit to users.
```

`Enabled by default`: True

`Autocorrects`: True (deletes the `if respond_to?` occurence)

#### Chef/CookbooksDependsOnSelf

The `CookbooksDependsOnSelf` cop detects a cookbook that depends on itself in metadata.rb.

`Enabled by default`: True

`Autocorrects`: False

#### Chef/MetadataMissingName

> The `MetadataMissingName` cop detects metadata.rb which is missing the name field required by Chef Infra Client 12+

Reviewer: ```suggestion
The `MetadataMissingName` cop detects a missing name field in the metadata.rb. Chef Infra Client 12+ requires the name field.
```

`Enabled by default`: True

`Autocorrects`: False

#### Chef/RequireRecipe

> The `RequireRecipe` cop detects the usage of `require_recipe` to include a recipe instead of `include_recipe`.

Reviewer: Is this required in all versions of Chef Infra Client, or just in more recent versions?

Author: for the last ~8 years I think. Basically forever

`Enabled by default`: True

`Autocorrects`: True

#### Chef/InvalidLicenseString

The `InvalidLicenseString` cop detects non-SPDX compliant license strings (or 'all rights reserved') in metadata.rb and autocorrects many common license typos.

`Enabled by default`: True

`Autocorrects`: True

#### Chef/SetOrReturnInResources

> The `SetOrReturnInResources` cop detects the usage of the `set_or_return` helper method in resources to create properties instead of using the `property` method. This was traditionally done when writing HWRPs, but properties created this way lack additional validation, reporting, and documentation functionality.

Reviewer: ```suggestion
The `SetOrReturnInResources` cop detects the usage of the `set_or_return` helper method in resources to create properties. Modern Chef Infra practice prefers using the `property` method because `set_or_return` lacks additional validation, reporting, and documentation functionality.
```

`Examples`

A resource property created using set_or_return:
```ruby
  def severity(arg = nil)
    set_or_return(
      :severity, arg,
      :kind_of => String,
      :default => nil
    )
  end
```

The same property created using the resource DSL
```ruby
  property :severity, String
```

`Enabled by default`: True

`Autocorrects`: False

### Other fixes and changes

> - The configuration file has been improved to properly verify LWRP/HWRP files

Reviewer: ```suggestion
- Improves the configuration file to correctly verify LWRP/HWRP files
```

> - Several Chef/* cops are now properly enabled in the config file

Reviewer: ```suggestion
- Enables several Chef/* cops in the config file
```

> - Chef/UseBuildEssentialResource is now enabled by default

Reviewer: ```suggestion
- Enables Chef/UseBuildEssentialResource by default
```

> - Chef/NodeSetUnless and Chef/NodeSet have been improved to properly correct usage of node.set in ChefSpec tests

Reviewer: ```suggestion
- Fixes Chef/NodeSetUnless and Chef/NodeSet to properly correctly use of node.set in ChefSpec tests
```
